spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindableProxyFactory.java

```diff
@@ -0,0 +1,316 @@
+	}
+
+	private MessageChannel createMessageChannel(Class<?> messageChannelType) {
+		return isPollable(messageChannelType) ? new QueueChannel() : new DirectChannel();
+	}
+
+	private boolean isPollable(Class<?> channelType) {
+		return PollableChannel.class.equals(channelType);
+	}
+
+	@Override
+	public synchronized Object invoke(MethodInvocation invocation) throws Throwable {
+		Method method = invocation.getMethod();
+		if (MessageChannel.class.isAssignableFrom(method.getReturnType())) {
+			Input input = AnnotationUtils.findAnnotation(method, Input.class);
+			if (input != null) {
+				String name = BindingBeanDefinitionRegistryUtils.getChannelName(input,
+						method);
+				return this.inputs.get(name).getMessageChannel();
+			}
+			Output output = AnnotationUtils.findAnnotation(method, Output.class);
+			if (output != null) {
+				String name = BindingBeanDefinitionRegistryUtils.getChannelName(output,
+						method);
+				return this.outputs.get(name).getMessageChannel();
+			}
+		}
+		// ignore
+		return null;
+	}
+
+	@Override
+	public synchronized Object getObject() throws Exception {
+		if (this.proxy == null) {
+			createChannels(this.type);
+			ProxyFactory factory = new ProxyFactory(type, this);
+			this.proxy = factory.getProxy();
+		}
+		return this.proxy;
+	}
+
+	@Override
+	public Class<?> getObjectType() {
+		return this.type;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	@Override
+	public void bindInputs(ChannelBindingAdapter channelBindingAdapter) {
+		for (Map.Entry<String, ChannelHolder> channelHolderEntry : inputs.entrySet()) {
+			ChannelHolder channelHolder = channelHolderEntry.getValue();
+			if (channelHolder.isBound()) {
+				channelBindingAdapter.bindMessageConsumer(
+						channelHolder.getMessageChannel(), channelHolderEntry.getKey());
+			}
+		}
+	}
+
+	@Override
+	public void bindOutputs(ChannelBindingAdapter channelBindingAdapter) {
+		for (Map.Entry<String, ChannelHolder> channelHolderEntry : outputs.entrySet()) {
+			if (channelHolderEntry.getValue().isBound()) {
+				channelBindingAdapter.bindMessageProducer(channelHolderEntry.getValue()
+						.getMessageChannel(), channelHolderEntry.getKey());
+			}
+		}
+	}
+
+	@Override
+	public void unbindInputs(ChannelBindingAdapter channelBindingAdapter) {
+		for (Map.Entry<String, ChannelHolder> channelHolderEntry : inputs.entrySet()) {
+			if (channelHolderEntry.getValue().isBound()) {
+				channelBindingAdapter.unbindConsumers(channelHolderEntry.getKey());
+			}
+		}
+	}
+
+	@Override
+	public void unbindOutputs(ChannelBindingAdapter channelBindingAdapter) {
+		for (Map.Entry<String, ChannelHolder> channelHolderEntry : outputs.entrySet()) {
+			if (channelHolderEntry.getValue().isBound()) {
+				channelBindingAdapter.unbindProducers(channelHolderEntry.getKey());
+			}
+		}
+	}
+
+	/**
+	 * Holds information about the channels exposed by the interface proxy, as well as
+	 * their status.
+	 *
+	 */
+	class ChannelHolder {
```

Comment:
This would at least save memory (reference to enclosing class), as well as protect from potential leaks. And it's just the right thing to do.